gax/src/main/java/com/google/api/gax/batching/FlowController.java

```diff
@@ -213,25 +234,106 @@ public void release(long elements, long bytes) {
       outstandingElementCount.release(elements);
     }
     if (outstandingByteCount != null) {
-      // Need to return at most as much bytes as it can be drawn.
-      long permitsToReturn = Math.min(bytes, maxOutstandingRequestBytes);
-      outstandingByteCount.release(permitsToReturn);
+      outstandingByteCount.release(bytes);
+    }
+  }
+
+  /**
+   * Increase flow control limits to allow extra elementSteps elements and byteSteps request bytes
+   * before enforcing flow control.
+   */
+  @InternalApi("For google-cloud-java client use only")
+  public void increaseThresholds(long elementSteps, long byteSteps) {
+    Preconditions.checkArgument(elementSteps >= 0);
+    Preconditions.checkArgument(byteSteps >= 0);
+    synchronized (updateLimitLock) {
+      if (outstandingElementCount != null) {
+        long actualStep = Math.min(elementSteps, maxElementCountLimit - currentElementCountLimit);
+        currentElementCountLimit += actualStep;
+        outstandingElementCount.release(actualStep);
+      }
+
+      if (outstandingByteCount != null) {
+        long actualStep = Math.min(byteSteps, maxRequestBytesLimit - currentRequestBytesLimit);
+        currentRequestBytesLimit += actualStep;
+        outstandingByteCount.release(actualStep);
+      }
+    }
+  }
+
+  /**
+   * Decrease flow control limits to allow elementSteps fewer elements and byteSteps fewer request
+   * bytes before enforcing flow control.
+   */
+  @InternalApi("For google-cloud-java client use only")
+  public void decreaseThresholds(long elementSteps, long byteSteps) {
+    Preconditions.checkArgument(elementSteps >= 0);
+    Preconditions.checkArgument(byteSteps >= 0);
+    synchronized (updateLimitLock) {
+      if (outstandingElementCount != null) {
+        long actualStep = Math.min(elementSteps, currentElementCountLimit - minElementCountLimit);
+        currentElementCountLimit -= actualStep;
```

Comment:
Please check the synchronization commment in `increaseThresholds()` method. The same applies here.